### ISSUE TITLE
Mipmap Generation

### ISSUE THREAD

**Issue by tetterl** [BachiLi/redner#95]
First of all thanks for implementing the mipmap genration!

While training a texture in tensorflow I noticed that the mipmap generation is a lot slower in TensorFlow compared to PyTorch (5x-20x depending on CPU/GPU).
See: https://colab.research.google.com/drive/1SOxiIwsPL3-fUsjrPamYIhdp5aWxod5P

In my current code the mipmap generation takes as long as the rendering process (multiple batches of textures, batch rendering). Do you have any idea how we could speed up the mipmap generation?

Furthermore, I don't understand current mipmap generation, which is:

1. Pad periodically
2. Depthwise convolution with box filter (smoothing to prevent aliasing)
3. Resize bilinear with antialiasing enabled.

Wouldn't be one of the following approaches be enough?

- Depthwise convolution with box filter and `strides=[1, 2, 2, 1]` (classical box filter downsamling), as shown in colab file
- resize with the current options only (without filtering)
This way we wouldn't do the antialiasing twice.

In the case of batched textures it might be reasonable to also have a batched mipmap generation

**Comment by BachiLi** [comment]
Padding is necessary for correctness since redner textures always wrap around. You might be able to resize by subsampling without losing much quality. Do you know which one is the bottleneck?

**Comment by BachiLi** [comment]
also switching to NCHW might be much faster than NHWC since cuDNN uses NCHW

**Comment by tetterl** [comment]
Thanks, I see.
According to the measurements below `resize` is the bottleneck and switching to NCHW doesn't improve the runtime. I have no clue why TensorFlow is so much slower.

```TensorFlow Texture: 0.8545131683349609
TensorFlow Box NHWC Filter: 0.11232900619506836
TensorFlow Box NCHW Filter: 0.12042760848999023
TensorFlow Resize Filter: 0.6337606906890869
PyTorch Texture: 0.05032157897949219```
(see colab file above)

If no change of the source code is possible to the wrapping around requirement I'll roll my own custom mipmap generation. In addition it might be worth to give the user the possibility to disable the the mipmap generation (only the base level in the mipmap list) if that doesn't break the other code.

**Comment by BachiLi** [comment]
Changing to your strided convoluton sounds good to me. A PR would be very helpful as I have other work to do. Otherwise I'll fix this two weeks from now.

**Comment by tetterl** [comment]
So without any padding then? The problematic thing is that this would only work (without cropping pixels of) for textures of size 2^n.
If that's ok I'll do a PR tomorrow.

**Comment by BachiLi** [comment]
Ah I forgot about padding. Padding is necessary. If you know how to do this with padding then a PR is welcome. Otherwise I'll come back to this after my deadline next Thursday.

**Comment by tetterl** [comment]
Since I don't have time right now to look at the underlying code (and why/how padding is necssary) I'll use my custom batched mipmap generation for the moment.

**Comment by BachiLi** [comment]
Regarding the necessity of the padding: imagine you are looking up position (50, 127) at an image with size (128, 128). Now you want to average the neighbor to prefilter the texture. What is at (50, 128)? It depends on the "texture warping mode". Redner's texture lookup always wrap around (so (50, 128) -> (50, 0)). Not doing this in the wrap-around mode usually results in a "seam" on the rendering where the edge of the texture meets.

**Comment by tetterl** [comment]
Added some ideas in the colab file (https://colab.research.google.com/drive/1SOxiIwsPL3-fUsjrPamYIhdp5aWxod5P, not tested). Might be an overkill to use a Gaussian 4x4 filter. I'm applying circular/wrap padding around the whole texture and then convolve with a 4x4 filter with a stride of 2. The idea was that this way the area of 2x2 texels maps to a 1x1 texel on the next level and the sizes will be a power of two if we start with one.
Since TensorFlow has a wrap/circular padding the whole mipmap generation is still not as fast as the current PyTorch version.

**Comment by BachiLi** [comment]
I switched from image.resize to 2d average pooling. Can you check if they are faster & produce desirable results?

**Comment by tetterl** [comment]
It's faster and produces the expected/desired results on my applications. Many thanks:

```
Old: 
TensorFlow Texture: 0.8545131683349609
New:
TensorFlow Texture: 0.238938570022583

PyTorch Texture: 0.05032157897949219
```
